tests/unit/type/stream-cgroups.tcl

```diff
@@ -205,11 +205,45 @@ start_server {
         $rd close
     }
 
+    test {Blocking XREADGROUP for stream that ran dry (issue #5299)} {
+        set rd [redis_deferring_client]
+
+        # Add a entry then delete it, now stream's last_id is 666.
+        r DEL mystream
+        r XGROUP CREATE mystream mygroup $ MKSTREAM
+        r XADD mystream 666 key value
+        r XDEL mystream 666
+
+        # Return right away instead of blocking, return the stream with an
+        # empty list instead of NIL if the ID specified is not the special `>` ID.
+        foreach id {0 600 666 700} {
+            $rd XREADGROUP GROUP mygroup myconsumer BLOCK 0 STREAMS mystream $id
+            assert_equal [$rd read] {{mystream {}}}
+        }
```

Comment:
i'm no expert to streams, but this looks odd to me.
why would it return with an empty stream and not block or nil?
@itamarhaber ?

Reply:
In https://redis.io/commands/XREADGROUP

The ID to specify in the STREAMS option when using XREADGROUP can be one of the following two:

- The special > ID, which means that the consumer want to receive only messages that were never delivered to any other consumer. It just means, give me new messages.
- Any other ID, that is, 0 or any other valid ID or incomplete ID (just the millisecond time part), will have the effect of returning entries that are pending for the consumer sending the command with IDs greater than the one provided. So basically if the ID is not >, then the command will just let the client access its pending entries: messages delivered to it, but not yet acknowledged. Note that in this case, both BLOCK and NOACK are ignored.

and in the code, the location is probably here: https://github.com/redis/redis/blob/unstable/src/t_stream.c#L2133

Reply:
If there are new messages in the stream, would it deliver them? 
If yes, then BLOCK can wait for them, if not, then why allow that BLOCK syntax?

Reply:
> If there are new messages in the stream, would it deliver them?

it will not. if ID is not `>`, it will always look up the pending list (non-ack entry list), find the elements that are smaller than the specified ID

> then why allow that BLOCK syntax

i don't know, look like if ID is not `>`, XREADGROUP BLOCK will always ignore BLOCK, I added a new test case to demonstrate (it has nothing to do with the original issue, but anyway, feel it is needed)

Reply:
I agree with @enjoy-binbin 

XREADGROUP works in two ways:

1. When given an ID, it will try to return it and up to COUNT entries from the consumer's pending (already delivered, never lacked) list. A null array means there are no such entries. It also ignores BLOCK, because the only way to add new PEL entries is via no. 2.
2. When given the special `>` ID, it will try to find new group messages (with IDs higher than the group's last delivered). The BLOCK is only effective in this case.